server/graphql/printer/mutations/insert.rs

```diff
@@ -0,0 +1,111 @@
+    simple_generic_errors::{RecordAlreadyExist, RecordNotFound},
+    standard_graphql_error::{validate_auth, StandardGraphqlError},
+    ContextExt,
+};
+use graphql_types::types::PrinterNode;
+use repository::PrinterRow;
+use service::{
+    auth::{Resource, ResourceAccessRequest},
+    printer::{InsertPrinter, InsertPrinterError},
+};
+#[derive(InputObject)]
+pub struct InsertPrinterInput {
+    pub id: String,
+    pub description: String,
+    pub address: String,
+    pub port: u16,
+    pub label_width: i32,
+    pub label_height: i32,
+}
+
+#[derive(SimpleObject)]
+pub struct InsertError {
+    pub error: InsertErrorInterface,
+}
+
+#[derive(Union)]
+pub enum InsertPrinterResponse {
+    Error(InsertError),
+    Response(PrinterNode),
+}
+
+#[derive(Interface)]
+#[graphql(name = "InsertPrinterErrorInterface")]
+#[graphql(field(name = "description", ty = "String"))]
+pub enum InsertErrorInterface {
+    DuplicatePrinter(RecordAlreadyExist),
+    CreatedRecordNotFound(RecordNotFound),
+}
+
+pub fn insert_printer(
+    ctx: &Context<'_>,
+    store_id: String,
+    input: InsertPrinterInput,
+) -> Result<InsertPrinterResponse> {
+    validate_auth(
+        ctx,
+        &ResourceAccessRequest {
+            resource: Resource::MutateItems,
+            store_id: Some(store_id.to_string()),
+        },
+    )?;
+
+    let service_provider = ctx.service_provider();
+    let service_context = service_provider.basic_context()?;
+
+    let result = service_provider
+        .printer_service
+        .insert_printer(&service_context, input.to_domain());
+    map_response(result)
+}
+
+impl InsertPrinterInput {
+    pub fn to_domain(self) -> InsertPrinter {
+        let InsertPrinterInput {
+            id,
+            description,
+            address,
+            port,
+            label_width,
+            label_height,
+        } = self;
+
+        InsertPrinter {
+            id,
+            description,
+            address,
+            port,
+            label_width,
+            label_height,
+        }
+    }
+}
+
+pub fn map_response(from: Result<PrinterRow, InsertPrinterError>) -> Result<InsertPrinterResponse> {
+    let result = match from {
+        Ok(response) => InsertPrinterResponse::Response(PrinterNode::from_domain(response)),
+
+        Err(error) => InsertPrinterResponse::Error(InsertError {
+            error: map_error(error)?,
+        }),
+    };
+
+    Ok(result)
+}
+
+fn map_error(error: InsertPrinterError) -> Result<InsertErrorInterface> {
+    use StandardGraphqlError::*;
+    let formatted_error = format!("{:#?}", error);
+
+    let graphql_error = match error {
+        InsertPrinterError::DuplicatePrinterId => BadUserInput(formatted_error),
+        InsertPrinterError::DuplicatePrinterDescription => BadUserInput(formatted_error),
+        InsertPrinterError::DuplicatePrinterAddress => BadUserInput(formatted_error),
+        InsertPrinterError::CreatedRecordNotFound => BadUserInput(formatted_error),
```

Comment:
created record not found would be an internal error (something went wrong with saving/finding the data, rather than the inputs being bad)

Reply:
just this one? 🙏